src/main/java/org/jboss/msc/service/ServiceControllerImpl.java

```diff
@@ -1621,9 +1593,11 @@ final boolean execute() {
     }
 
     private abstract class DependentsControllerTask extends ControllerTask {
-        protected final Dependent[][] dependents;
+        protected final Map<ServiceName, Dependent[]> dependents;
+        protected final Dependent[] children;
         private DependentsControllerTask() {
-            dependents = getDependents();
+            dependents = getDependentsByDependencyName();
+            children = ServiceControllerImpl.this.children.toScatteredArray(NO_DEPENDENTS);
```

Comment:
There is no need for benchmarking now because this is work in progress that will be changed in next pull request.